Adds `startup_probe` field to `google_cloud_run_service` resource for beta

yanweiguo opened the pull request:
<!-- Put a description of what this PR is for here, along with any references to issues that this resolves or contributes to -->

Part of https://github.com/hashicorp/terraform-provider-google/issues/12532.

Adds `startup_probe` field to `google_cloud_run_service` resource for google beta provider.



<!--
Replace each [ ] with [X] to check it. Switch to the preview view to make it easier to click on links.
These steps will speed up the review process, and we appreciate you spending time on them before sending
your code to be reviewed.
-->
If this PR is for Terraform, I acknowledge that I have:

- [X] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-the-terraform-providers), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [x] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/main/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [x] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

<!-- AUTOCHANGELOG for Downstream PRs.

Please select one of the following "release-note:" headings:
    - release-note:enhancement
    - release-note:bug
    - release-note:note
    - release-note:new-resource
    - release-note:new-datasource
    - release-note:deprecation
    - release-note:breaking-change
    - release-note:none
    
Unless you choose release-note:none, please add a release note.

See .ci/RELEASE_NOTES_GUIDE.md for writing good release notes.

You can add more release note blocks if you want more than one CHANGELOG
entry for this PR.
-->
**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
cloudrun: added field `startup_probe` to resource `google_cloud_run_service` (beta)
```

modular-magician commented:
Hello!  I am a robot who works on Magic Modules PRs.

I've detected that you're a community contributor. @rileykarson, a repository maintainer, has been assigned to assist you and help review your changes. 

<details>
  <summary>:question: First time contributing? Click here for more details</summary>

---

Your assigned reviewer will help review your code by: 
* Ensuring it's backwards compatible, covers common error cases, etc.
* Summarizing the change into a user-facing changelog note.
* Passes tests, either our "VCR" suite, a set of presubmit tests, or with manual test runs.

You can help make sure that review is quick by running local tests and ensuring they're passing in between each push you make to your PR's branch. Also, try to leave a comment with each push you make, as pushes generally don't generate emails.

If your reviewer doesn't get back to you within a week after your most recent change, please feel free to leave a comment on the issue asking them to take a look! In the absence of a dedicated review dashboard most maintainers manage their pending reviews through email, and those will sometimes get lost in their inbox.

---

</details>

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-6532-old..auto-pr-6532) ( 1 file changed, 70 insertions(+))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-6532-old..auto-pr-6532) ( 3 files changed, 624 insertions(+))
TF Validator: [Diff](https://github.com/modular-magician/terraform-validator/compare/auto-pr-6532-old..auto-pr-6532) ( 2 files changed, 3 insertions(+), 3 deletions(-))

modular-magician commented:
#### Tests analytics 
Total tests: `2171` 
Passed tests `1931` 
Skipped tests: `236` 
Failed tests: `4` 

#### Action taken 
<details> <summary>Triggering VCR tests in RECORDING mode for the tests that failed during VCR. Click here to see the failed tests</summary><blockquote>TestAccPrivatecaCertificateAuthority_privatecaCertificateAuthoritySubordinateExample|TestAccComputeInstance_soleTenantNodeAffinities|TestAccCloudRunService_probes|TestAccCGCSnippet_eventarcWorkflowsExample </blockquote></details>

modular-magician commented:
Tests passed during RECORDING mode:
 `TestAccPrivatecaCertificateAuthority_privatecaCertificateAuthoritySubordinateExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/769e6f1b-342e-414e-a5a2-8651e93440e1/recording/TestAccPrivatecaCertificateAuthority_privatecaCertificateAuthoritySubordinateExample.log)]
`TestAccCloudRunService_probes`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/769e6f1b-342e-414e-a5a2-8651e93440e1/recording/TestAccCloudRunService_probes.log)]
`TestAccCGCSnippet_eventarcWorkflowsExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/769e6f1b-342e-414e-a5a2-8651e93440e1/recording/TestAccCGCSnippet_eventarcWorkflowsExample.log)] 

Tests failed during RECORDING mode:
 `TestAccComputeInstance_soleTenantNodeAffinities`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/769e6f1b-342e-414e-a5a2-8651e93440e1/build-log/recording_build/TestAccComputeInstance_soleTenantNodeAffinities_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/769e6f1b-342e-414e-a5a2-8651e93440e1/recording/TestAccComputeInstance_soleTenantNodeAffinities.log)] 

Please fix these to complete your PR
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/769e6f1b-342e-414e-a5a2-8651e93440e1/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/769e6f1b-342e-414e-a5a2-8651e93440e1/recording) for each test

yanweiguo commented:
I don't think the failed test is relative to my changes.

rileykarson commented on mmv1/products/cloudrun/api.yaml:
```diff
@@ -616,6 +616,68 @@ objects:
+                    TcpSocket speficies an action involving a TCP port.
+                  properties:
+                    - !ruby/object:Api::Type::Integer
+                      name: port
```
Should this field be made required? Otherwise, it will be possible to specify the following, likely resulting in an error:

```
tcp_socket {}
```

yanweiguo replied:
```
tcp_socket {}
```
is a valid input.

yanweiguo replied:
In our API, port is optional. Default value will be set for port if it is not specified.

However I tested with `tcp_socket {}` and got an error saying either tcp_scoket or http_get should be set. Does terraform ignore tcp_socket field if it is empty?

yanweiguo replied:
If a user doesn't specify `tcp_socket`, the API won't return it.

The default value is equal to `container.ports[0].port`, which is input by user or defaults to `8080`.

rileykarson commented on mmv1/products/cloudrun/api.yaml:
```diff
@@ -616,6 +616,68 @@ objects:
+                    HttpGet specifies the http request to perform..
+                  properties:
+                    - !ruby/object:Api::Type::String
+                      name: path
```
Can this value ever be empty? It seems likely that it should be required.

yanweiguo replied:
Added a test to cover it.

rileykarson replied:
We don't appear to be updating the field in the test- see the comment about `default_from_api` on the field, we likely need a `default_value` instead.

I think Terraform Core's eating the diff there- it's considering `""` equal to unset and preserving the old value.

yanweiguo replied:
I don't think it's considering "" equal to unset. I tried `default_value = "/"`. The test with `path = ""` failed.

rileykarson replied:
Ah! I see what's happening- we're sending the following in the update:

```
"startupProbe": {
       "failureThreshold": 3,
       "httpGet": {
        "httpHeaders": [
         {
          "name": "User-Agent",
          "value": "magic-modules"
         }
        ]
       },
       "periodSeconds": 10,
       "timeoutSeconds": 1
      }
```

Notice that `path` isn't present. While this is a PUT, and per normal PUT rules we'd expect lack of presence to mean the type's empty value, this isn't interpreted that way by the API. It's either preserving the value or setting the default value instead, likely preserving it. That means that we see the following on return:

```
            "startupProbe": {
              "timeoutSeconds": 1,
              "periodSeconds": 10,
              "failureThreshold": 3,
              "httpGet": {
                "path": "/",
                "httpHeaders": [
                  {
                    "name": "User-Agent",
                    "value": "magic-modules"
                  }
                ]
              }
```

We can resolve that by setting `send_empty_value: true` on the field, as described at https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1#apiyaml

rileykarson commented on mmv1/products/cloudrun/api.yaml:
```diff
@@ -616,6 +616,68 @@ objects:
+                      item_type: !ruby/object:Api::Type::NestedObject
+                        properties:
+                          - !ruby/object:Api::Type::String
+                            name: name
```
Same as above with both these subfields

yanweiguo replied:
I'm not sure why `http_headers {}` causes test failing but not `http_headers {name = "somename"}` or `http_headers {value = "somevalue"}

rileykarson replied:
If that causes a test failure we should cover it with a test & work out how to get it to pass.

Is this a valid specification?

```
http_headers {
  name = ""
}
```

yanweiguo replied:
Yes it's a valid specification. I tried a test for it but it failed due to

```
           ~ http_get {
                  # (1 unchanged attribute hidden)

                + http_headers {}

                  # (1 unchanged block hidden)
              }
```

Do you have any suggestions to fix it?

rileykarson replied:
This may need `send_empty_value: true` as well, per above!

rileykarson commented on mmv1/products/cloudrun/api.yaml:
```diff
@@ -616,6 +616,68 @@ objects:
+                  description: |-
+                    Number of seconds after the container has started before the probe is
+                    initiated.
+                    Defaults to 0 seconds. Minimum value is 0. Maximum value is 240.
```
I see that for these integer fields, you've set them as `default_from_api` rather than setting clientside defaults. Is that for a specific reason? Assuming these values are fixed in the API, Terraform-local defaults would be preferable.

See https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/README.md#terraformyaml for details on `default_from_api`. It's a good solution when necessary, but pretty heavyweight- local defaults are significantly better when possible.

yanweiguo replied:
Didn't know that `default_value` is available. Done.

rileykarson commented on mmv1/products/cloudrun/terraform.yaml:
```diff
@@ -390,6 +390,20 @@ overrides: !ruby/object:Overrides::ResourceOverrides
         default_from_api: true
       spec.template.spec.containers.resources.limits: !ruby/object:Overrides::Terraform::PropertyOverride
         default_from_api: true
+      spec.template.spec.containers.startupProbe: !ruby/object:Overrides::Terraform::PropertyOverride
```
This doesn't need `default_from_api`, as the field doesn't appear to be set by default in the API.

yanweiguo replied:
We will set default startupProbe soon.

yanweiguo replied:
ack

rileykarson commented on mmv1/templates/terraform/examples/cloud_run_service_probes.tf.erb:
```diff
@@ -0,0 +1,35 @@
+data "google_project" "project" {
```
This value isn't referenced anywhere- you can remove the datasource

rileykarson commented on mmv1/templates/terraform/examples/cloud_run_service_probes.tf.erb:
```diff
@@ -0,0 +1,35 @@
+  provider = google-beta
+}
+
+resource "google_cloud_run_service" "<%= ctx[:primary_resource_id] %>" {
```
In `terraform.yaml` you'll need to add an `examples` entry to register this! https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/README.md#terraformyaml-metadata

yanweiguo replied:
Done.

rileykarson commented on mmv1/third_party/terraform/tests/resource_cloud_run_service_test.go.erb:
```diff
@@ -359,3 +359,132 @@ resource "google_cloud_run_service" "default" {
+
+<% unless version == 'ga' -%>
+
+// this test checks that Terraform does not fail with a 409 recreating the same service
```
```suggestion
```

rileykarson reviewed (commented):
Made a first pass! I have a few more thoughts about what we want to test, but those are contingent on how we handle default values for the integer fields.

rileykarson commented:
> I don't think the failed test is relative to my changes.

Yep, you're all good. Sorry for the noise!

rileykarson commented on mmv1/products/cloudrun/terraform.yaml:
```diff
@@ -332,6 +332,14 @@ overrides: !ruby/object:Overrides::ResourceOverrides
         vars:
           cloud_run_srv: "cloud-run-srv"
+      - !ruby/object:Provider::Terraform::Examples
+        name: "cloud_run_service_probes"
```
We'll need to mark this as a beta test per https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/README.md#tests-that-use-a-beta-feature (including annotating the example code)

That's currently blocking generation. It shouldn't, but there's a piece of CI making a bad assumption atm.

rileykarson commented on mmv1/products/cloudrun/terraform.yaml:
```diff
@@ -390,6 +398,12 @@ overrides: !ruby/object:Overrides::ResourceOverrides
+        default_from_api: true
+      spec.template.spec.containers.startupProbe.tcpSocket.port: !ruby/object:Overrides::Terraform::PropertyOverride
+        default_from_api: true
+      spec.template.spec.containers.startupProbe.httpGet.path: !ruby/object:Overrides::Terraform::PropertyOverride
```
This indicates that the server will pick a non-fixed default for path when not provided. Is that correct?

yanweiguo replied:
The rules are:

- Not specified -> "/"
- Empty string -> "/"
- Other string -> keep the input

rileykarson replied:
In that case, should we consider a client-side `default_value` or `"/"`? That means that if a user doesn't specify anything, they'll get `"/"` which is the same behaviour as the API.

`default_from_api` preserves values when the field is unset rather than resets to default, so it's generally undesirable if the default value is fixed.

yanweiguo replied:
I tried `default_value = "/"`.

If I don't specify `path`, the test passed.

If I specify `path=""`, the test failed due to

```
~ http_get {
        - path = "/" -> null

          # (3 unchanged blocks hidden)
      }
```

I think Terraform treats "" as an input, not unset.

rileykarson reviewed (commented):
Looks like generation got blocked, that's a bug on our end. Details on how to resolve inline (along with some more implementation questions I was hoping to answer myself from the test logs post-generation. Sorry for the extra qs!)

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-6532-old..auto-pr-6532) ( 1 file changed, 108 insertions(+))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-6532-old..auto-pr-6532) ( 4 files changed, 740 insertions(+))
TF Validator: [Diff](https://github.com/modular-magician/terraform-validator/compare/auto-pr-6532-old..auto-pr-6532) ( 2 files changed, 3 insertions(+), 3 deletions(-))

modular-magician commented:
#### Tests analytics 
Total tests: `2173` 
Passed tests `1929` 
Skipped tests: `238` 
Failed tests: `6` 

#### Action taken 
<details> <summary>Triggering VCR tests in RECORDING mode for the tests that failed during VCR. Click here to see the failed tests</summary><blockquote>TestAccComputeInstance_soleTenantNodeAffinities|TestAccCloudRunService_cloudRunServiceProbesExample|TestAccCloudRunService_probes|TestAccComputeForwardingRule_update|TestAccCGCSnippet_eventarcWorkflowsExample|TestAccFirebaserulesRelease_BasicRelease </blockquote></details>

modular-magician commented:
Tests passed during RECORDING mode:
 `TestAccCloudRunService_probes`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/feabee0a-cf8e-4d35-8e91-6a5a5b0ee3ad/recording/TestAccCloudRunService_probes.log)]
`TestAccComputeForwardingRule_update`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/feabee0a-cf8e-4d35-8e91-6a5a5b0ee3ad/recording/TestAccComputeForwardingRule_update.log)]
`TestAccFirebaserulesRelease_BasicRelease`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/feabee0a-cf8e-4d35-8e91-6a5a5b0ee3ad/recording/TestAccFirebaserulesRelease_BasicRelease.log)] 

Tests failed during RECORDING mode:
 `TestAccComputeInstance_soleTenantNodeAffinities`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/feabee0a-cf8e-4d35-8e91-6a5a5b0ee3ad/build-log/recording_build/TestAccComputeInstance_soleTenantNodeAffinities_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/feabee0a-cf8e-4d35-8e91-6a5a5b0ee3ad/recording/TestAccComputeInstance_soleTenantNodeAffinities.log)]
`TestAccCloudRunService_cloudRunServiceProbesExample`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/feabee0a-cf8e-4d35-8e91-6a5a5b0ee3ad/build-log/recording_build/TestAccCloudRunService_cloudRunServiceProbesExample_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/feabee0a-cf8e-4d35-8e91-6a5a5b0ee3ad/recording/TestAccCloudRunService_cloudRunServiceProbesExample.log)]
`TestAccCGCSnippet_eventarcWorkflowsExample`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/feabee0a-cf8e-4d35-8e91-6a5a5b0ee3ad/build-log/recording_build/TestAccCGCSnippet_eventarcWorkflowsExample_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/feabee0a-cf8e-4d35-8e91-6a5a5b0ee3ad/recording/TestAccCGCSnippet_eventarcWorkflowsExample.log)] 

Please fix these to complete your PR
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/feabee0a-cf8e-4d35-8e91-6a5a5b0ee3ad/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/feabee0a-cf8e-4d35-8e91-6a5a5b0ee3ad/recording) for each test

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-6532-old..auto-pr-6532) ( 1 file changed, 108 insertions(+))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-6532-old..auto-pr-6532) ( 4 files changed, 740 insertions(+))
TF Validator: [Diff](https://github.com/modular-magician/terraform-validator/compare/auto-pr-6532-old..auto-pr-6532) ( 2 files changed, 3 insertions(+), 3 deletions(-))

modular-magician commented:
#### Tests analytics 
Total tests: `2173` 
Passed tests `1932` 
Skipped tests: `238` 
Failed tests: `3` 

#### Action taken 
<details> <summary>Triggering VCR tests in RECORDING mode for the tests that failed during VCR. Click here to see the failed tests</summary><blockquote>TestAccFirebaserulesRelease_BasicRelease|TestAccComputeInstance_soleTenantNodeAffinities|TestAccCloudRunService_cloudRunServiceProbesExample </blockquote></details>

modular-magician commented:
Tests passed during RECORDING mode:
 `TestAccFirebaserulesRelease_BasicRelease`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/2fadd96e-fb38-4649-939a-74f5e27852b7/recording/TestAccFirebaserulesRelease_BasicRelease.log)] 

Tests failed during RECORDING mode:
 `TestAccComputeInstance_soleTenantNodeAffinities`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/2fadd96e-fb38-4649-939a-74f5e27852b7/build-log/recording_build/TestAccComputeInstance_soleTenantNodeAffinities_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/2fadd96e-fb38-4649-939a-74f5e27852b7/recording/TestAccComputeInstance_soleTenantNodeAffinities.log)]
`TestAccCloudRunService_cloudRunServiceProbesExample`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/2fadd96e-fb38-4649-939a-74f5e27852b7/build-log/recording_build/TestAccCloudRunService_cloudRunServiceProbesExample_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/2fadd96e-fb38-4649-939a-74f5e27852b7/recording/TestAccCloudRunService_cloudRunServiceProbesExample.log)] 

Please fix these to complete your PR
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/2fadd96e-fb38-4649-939a-74f5e27852b7/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/2fadd96e-fb38-4649-939a-74f5e27852b7/recording) for each test

yanweiguo commented:
For the failed [run](https://00f74ba44b2ee6b8ad02640cfa2371ad1ee272fae8-apidata.googleusercontent.com/download/storage/v1/b/ci-vcr-logs/o/beta%2Frefs%2Fheads%2Fauto-pr-6532%2Fartifacts%2F2fadd96e-fb38-4649-939a-74f5e27852b7%2Fbuild-log%2Frecording_build%2FTestAccCloudRunService_cloudRunServiceProbesExample_recording_test.log?jk=AFshE3UDAv3ALaTAwgccztqTB2TtUPoguUXBfhVoEyGJ-iNMXAPE_e29od2AfPMbGPCeTZrkrMiCtO_0XNR9tnPmACMAKoUpKHKQvXfnv4HESPpKHgjtimHKaQJz4dLxTRKdOl9lpufWVib0yWQj93Pkg67h69eRNesldyOAxu4TvSJ_GN4yrwa8kiCcXSlN5CZurkzWm8j9-gBIXKH3DMnuM23gm5a3a_XthloMXbsvsC9WBe6NYWOF0WvqcHnSUt_IHmBv8MXSdpD8iFJOevkfkJ1UOHr9YcsNZftDXeqrbDHRks9MtRw3QF4A-qj0VtVSGVF1PVsB9922naDdlX-BHZjCzu_FmNy0EQIcorxC0wsNB2JTMlGW8eiQy6re9MXTxGDvzB4x9sKurXGIZfsLmwv5EC0xVB1sqyICsWn3rJ2xpkBdOwTpgPIeJNhjb1EKJ0hyM9sSRTdVhdbNPPncAIIzNyujFSYb7o38xXxkdTk-ZX_lqtvvzwlIhM4rojNzM34hxSPTTCP1R4K3L-1nn7OknfQE7mqv0MvC2k_rmiZQU3lktlF-hGbL0jrppVFi5-t3DWIfDYbwvGrHnh3G0r6PJAXRSh2hTphCHftSQuSon0ivyr-z-j7eKi3rucqk_pFLwpUaiUs9-2a6eHEMGNik2fIVz9tCAZ-Jm0-G1L8iFD5ogaEunAp93PYth0aIbzuXWIeI4mmTg2jDq8bgxPWp14iNhqK-z1LWscQNk9fYuDt6WlPYO8qokO6PT3FhaLZQE9dLW9_eUnuFAHAVmV1zlkkZxBg9yZ_quVClWzG0hZ35sZEeqDG3XrxB6XzZ30uIaAHptwWmbW4MBBNJQLX7rq7gTXpD2croHYo2iwaA3y-n7CZmerwmG-8r0EAyjkafCS5b84JZLepCPHiV1oUvbEQtR6NgKdDWb8hHbDXzjVUnXNG7CT_FFkim5sBQaVE8P-v2IuOBC9Nhnmf2Cg1BU7RUG2Xgwip-2dsZ0bnAsmRMiFwg3N8W8KZ4RLdunfZ83PcC6vhehksK8TCEtcgvq50QIoAd90HHil5wR7peQliTfZ6P9JcpmN569oGKH4Ar3T-0R1-XICH4lDyh_R5vHn-2KAB5UOymPdGaVLkCvzmQGCdvCkCRICNnJT4BghybhB3QCue846OxVrSgrHkI-TKRtsKfEHjzVHJcUcrn6Y-E35sjkcPwpV9i-9rzRcKmweoBkIgSbXg&isca=1) of TestAccCloudRunService_cloudRunServiceProbesExample, it is expected that `run.googleapis.com/ingress` is added by the API.

I'm not sure why other examples tests succeeded. Shall I update `cloudRunGoogleProvidedAnnotations` to make `run.googleapis.com/ingress` an allowed annotation? Shall I skip test for the example since its case is covered by other tests I added.

rileykarson commented:
Mixed-managed fields like `annotations` have never worked entirely right. I'm confused why it's failing how it does- the `lifecycle.ignore_changes` should have prevented that. Are we able to add the value to the `annotations` block?

rileykarson reviewed (commented):
Sorry for the delay in getting back! It takes a bit to work through these empty/nil issues, and I've been pretty underwater. Some comments inline.

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-6532-old..auto-pr-6532) ( 1 file changed, 108 insertions(+))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-6532-old..auto-pr-6532) ( 4 files changed, 740 insertions(+))
TF Validator: [Diff](https://github.com/modular-magician/terraform-validator/compare/auto-pr-6532-old..auto-pr-6532) ( 2 files changed, 3 insertions(+), 3 deletions(-))

modular-magician commented:
#### Tests analytics 
Total tests: `2183` 
Passed tests `1936` 
Skipped tests: `238` 
Failed tests: `9` 

#### Action taken 
<details> <summary>Triggering VCR tests in RECORDING mode for the tests that failed during VCR. Click here to see the failed tests</summary><blockquote>TestAccStorageTransferJob_transferOptions|TestAccFirebaserulesRelease_BasicRelease|TestAccComputeInstance_soleTenantNodeAffinities|TestAccCloudRunService_cloudRunServiceProbesExample|TestAccPrivatecaCaPoolIamPolicyGenerated_withCondition|TestAccComputeRegionBackendServiceIamPolicyGenerated_withCondition|TestAccComputeMachineImageIamPolicyGenerated_withCondition|TestAccComputeInstanceIamPolicyGenerated_withCondition|TestAccComputeBackendServiceIamPolicyGenerated_withCondition </blockquote></details>

modular-magician commented:
Tests passed during RECORDING mode:
 `TestAccStorageTransferJob_transferOptions`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/35c7b2db-f864-4aea-ba07-5804f05dfd49/recording/TestAccStorageTransferJob_transferOptions.log)]
`TestAccFirebaserulesRelease_BasicRelease`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/35c7b2db-f864-4aea-ba07-5804f05dfd49/recording/TestAccFirebaserulesRelease_BasicRelease.log)]
`TestAccPrivatecaCaPoolIamPolicyGenerated_withCondition`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/35c7b2db-f864-4aea-ba07-5804f05dfd49/recording/TestAccPrivatecaCaPoolIamPolicyGenerated_withCondition.log)]
`TestAccComputeRegionBackendServiceIamPolicyGenerated_withCondition`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/35c7b2db-f864-4aea-ba07-5804f05dfd49/recording/TestAccComputeRegionBackendServiceIamPolicyGenerated_withCondition.log)]
`TestAccComputeMachineImageIamPolicyGenerated_withCondition`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/35c7b2db-f864-4aea-ba07-5804f05dfd49/recording/TestAccComputeMachineImageIamPolicyGenerated_withCondition.log)]
`TestAccComputeInstanceIamPolicyGenerated_withCondition`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/35c7b2db-f864-4aea-ba07-5804f05dfd49/recording/TestAccComputeInstanceIamPolicyGenerated_withCondition.log)]
`TestAccComputeBackendServiceIamPolicyGenerated_withCondition`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/35c7b2db-f864-4aea-ba07-5804f05dfd49/recording/TestAccComputeBackendServiceIamPolicyGenerated_withCondition.log)] 

Tests failed during RECORDING mode:
 `TestAccComputeInstance_soleTenantNodeAffinities`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/35c7b2db-f864-4aea-ba07-5804f05dfd49/build-log/recording_build/TestAccComputeInstance_soleTenantNodeAffinities_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/35c7b2db-f864-4aea-ba07-5804f05dfd49/recording/TestAccComputeInstance_soleTenantNodeAffinities.log)]
`TestAccCloudRunService_cloudRunServiceProbesExample`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/35c7b2db-f864-4aea-ba07-5804f05dfd49/build-log/recording_build/TestAccCloudRunService_cloudRunServiceProbesExample_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/35c7b2db-f864-4aea-ba07-5804f05dfd49/recording/TestAccCloudRunService_cloudRunServiceProbesExample.log)] 

Please fix these to complete your PR
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/35c7b2db-f864-4aea-ba07-5804f05dfd49/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/35c7b2db-f864-4aea-ba07-5804f05dfd49/recording) for each test

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-6532-old..auto-pr-6532) ( 1 file changed, 114 insertions(+))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-6532-old..auto-pr-6532) ( 4 files changed, 752 insertions(+))
TF Validator: [Diff](https://github.com/modular-magician/terraform-validator/compare/auto-pr-6532-old..auto-pr-6532) ( 2 files changed, 3 insertions(+), 3 deletions(-))

modular-magician commented:
#### Tests analytics 
Total tests: `2184` 
Passed tests `1943` 
Skipped tests: `238` 
Failed tests: `3` 

#### Action taken 
<details> <summary>Triggering VCR tests in RECORDING mode for the tests that failed during VCR. Click here to see the failed tests</summary><blockquote>TestAccComputeInstance_soleTenantNodeAffinities|TestAccCloudRunService_cloudRunServiceProbesExample|TestAccFirebaserulesRelease_BasicRelease </blockquote></details>

modular-magician commented:
Tests passed during RECORDING mode:
 `TestAccCloudRunService_cloudRunServiceProbesExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/12c621d9-3180-40de-9472-9465b3da5fe2/recording/TestAccCloudRunService_cloudRunServiceProbesExample.log)]
`TestAccFirebaserulesRelease_BasicRelease`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/12c621d9-3180-40de-9472-9465b3da5fe2/recording/TestAccFirebaserulesRelease_BasicRelease.log)] 

Tests failed during RECORDING mode:
 `TestAccComputeInstance_soleTenantNodeAffinities`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/12c621d9-3180-40de-9472-9465b3da5fe2/build-log/recording_build/TestAccComputeInstance_soleTenantNodeAffinities_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/12c621d9-3180-40de-9472-9465b3da5fe2/recording/TestAccComputeInstance_soleTenantNodeAffinities.log)] 

Please fix these to complete your PR
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/12c621d9-3180-40de-9472-9465b3da5fe2/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/12c621d9-3180-40de-9472-9465b3da5fe2/recording) for each test

darkangel01111991 commented:
What do you think about including the liveness_probe in this PR as suggested in the related opened issue?

yanweiguo commented:
> What do you think about including the liveness_probe in this PR as suggested in the related opened issue?

Talked to the reviewer. Liveness probe is preferred to be added in a separate PR.

modular-magician commented:
#### Tests analytics 
Total tests: `2194` 
Passed tests `1940` 
Skipped tests: `240` 
Failed tests: `14` 

#### Action taken 
<details> <summary>Triggering VCR tests in RECORDING mode for the tests that failed during VCR. Click here to see the failed tests</summary><blockquote>TestAccFirebaserulesRelease_BasicRelease|TestAccComputeVpnTunnel_vpnTunnelBetaExample|TestAccComputeRouterInterface_basic|TestAccComputeInstance_soleTenantNodeAffinities|TestAccComputeForwardingRule_networkTier|TestAccComputeForwardingRule_forwardingRuleRegionalHttpXlbExample|TestAccComputeGlobalForwardingRule_externalTcpProxyLbMigBackendExample|TestAccComputeForwardingRule_forwardingRuleExternallbExample|TestAccComputeForwardingRule_internalTcpUdpLbWithMigBackendExample|TestAccClouddeployDeliveryPipeline_DeliveryPipeline|TestAccCloudRunService_probes|TestAccSqlDatabaseInstance_mysqlMajorVersionUpgrade|TestAccComputeFirewallPolicyRule_update|TestAccComputeFirewallPolicy_update </blockquote></details>

modular-magician commented:
Tests passed during RECORDING mode:
 `TestAccFirebaserulesRelease_BasicRelease`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/f4062062-1957-4844-a2b4-e14dd2932344/recording/TestAccFirebaserulesRelease_BasicRelease.log)]
`TestAccComputeVpnTunnel_vpnTunnelBetaExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/f4062062-1957-4844-a2b4-e14dd2932344/recording/TestAccComputeVpnTunnel_vpnTunnelBetaExample.log)]
`TestAccComputeRouterInterface_basic`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/f4062062-1957-4844-a2b4-e14dd2932344/recording/TestAccComputeRouterInterface_basic.log)]
`TestAccComputeForwardingRule_networkTier`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/f4062062-1957-4844-a2b4-e14dd2932344/recording/TestAccComputeForwardingRule_networkTier.log)]
`TestAccComputeForwardingRule_forwardingRuleRegionalHttpXlbExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/f4062062-1957-4844-a2b4-e14dd2932344/recording/TestAccComputeForwardingRule_forwardingRuleRegionalHttpXlbExample.log)]
`TestAccComputeGlobalForwardingRule_externalTcpProxyLbMigBackendExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/f4062062-1957-4844-a2b4-e14dd2932344/recording/TestAccComputeGlobalForwardingRule_externalTcpProxyLbMigBackendExample.log)]
`TestAccComputeForwardingRule_forwardingRuleExternallbExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/f4062062-1957-4844-a2b4-e14dd2932344/recording/TestAccComputeForwardingRule_forwardingRuleExternallbExample.log)]
`TestAccComputeForwardingRule_internalTcpUdpLbWithMigBackendExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/f4062062-1957-4844-a2b4-e14dd2932344/recording/TestAccComputeForwardingRule_internalTcpUdpLbWithMigBackendExample.log)]
`TestAccClouddeployDeliveryPipeline_DeliveryPipeline`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/f4062062-1957-4844-a2b4-e14dd2932344/recording/TestAccClouddeployDeliveryPipeline_DeliveryPipeline.log)]
`TestAccSqlDatabaseInstance_mysqlMajorVersionUpgrade`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/f4062062-1957-4844-a2b4-e14dd2932344/recording/TestAccSqlDatabaseInstance_mysqlMajorVersionUpgrade.log)]
`TestAccComputeFirewallPolicyRule_update`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/f4062062-1957-4844-a2b4-e14dd2932344/recording/TestAccComputeFirewallPolicyRule_update.log)]
`TestAccComputeFirewallPolicy_update`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/f4062062-1957-4844-a2b4-e14dd2932344/recording/TestAccComputeFirewallPolicy_update.log)] 

Tests failed during RECORDING mode:
 `TestAccComputeInstance_soleTenantNodeAffinities`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/f4062062-1957-4844-a2b4-e14dd2932344/build-log/recording_build/TestAccComputeInstance_soleTenantNodeAffinities_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/f4062062-1957-4844-a2b4-e14dd2932344/recording/TestAccComputeInstance_soleTenantNodeAffinities.log)]
`TestAccCloudRunService_probes`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/f4062062-1957-4844-a2b4-e14dd2932344/build-log/recording_build/TestAccCloudRunService_probes_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/f4062062-1957-4844-a2b4-e14dd2932344/recording/TestAccCloudRunService_probes.log)] 

Please fix these to complete your PR
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/f4062062-1957-4844-a2b4-e14dd2932344/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/f4062062-1957-4844-a2b4-e14dd2932344/recording) for each test

yanweiguo commented:
Per offline discussion, `path = ""` can be treated invalided in terraform (not  in API). I added a comment for that.

I have changed the tests to cover the following cases:

1. 
```
startupProbe {
  tcp {}
}
```
Note: some default values are auto added in the request set by terraform.

2. 
```
"startupProbe": {
   "failureThreshold": 2,
   "initialDelaySeconds": 2,
   "periodSeconds": 5,
   "tcpSocket": {
    "port": 8080
 }
```

3. 
```
startupProbe {
  httpGet {}
}
```
Note: some default values are auto added in the request set by terraform, including `httpGet.path = "/"`

4.
```
"startupProbe": {
 "failureThreshold": 3,
 "httpGet": {
  "httpHeaders": [
   {
    "name": "User-Agent",
    "value": "magic-modules"
   },
   {
    "name": "Some-Name",
    "value": ""
   },
   {
    "name": "",
    "value": "Some-Value"
   }
  ],
  "path": "/some-path"
 },
```

I still couldn't make the following case pass the test even thought with `allow_empty_object: true` and `send_empty_value: true`:

```
  http_headers {} or   http_headers { name = "" }
```

However I don't worry about that since the case that the name and value of a header are both empty really doesn't make sense.

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-6532-old..auto-pr-6532) ( 1 file changed, 114 insertions(+))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-6532-old..auto-pr-6532) ( 4 files changed, 825 insertions(+))
TF Validator: [Diff](https://github.com/modular-magician/terraform-validator/compare/auto-pr-6532-old..auto-pr-6532) ( 2 files changed, 3 insertions(+), 3 deletions(-))

modular-magician commented:
#### Tests analytics 
Total tests: `2195` 
Passed tests `1949` 
Skipped tests: `240` 
Failed tests: `6` 

#### Action taken 
<details> <summary>Triggering VCR tests in RECORDING mode for the tests that failed during VCR. Click here to see the failed tests</summary><blockquote>TestAccFirebaserulesRelease_BasicRelease|TestAccComputeInstance_soleTenantNodeAffinities|TestAccCloudRunService_probes|TestAccCloudRunService_cloudRunServiceProbesExample|TestAccCGCSnippet_eventarcWorkflowsExample|TestAccSqlDatabaseInstance_mysqlMajorVersionUpgrade </blockquote></details>

modular-magician commented:
Tests passed during RECORDING mode:
 `TestAccFirebaserulesRelease_BasicRelease`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/cca5eaa9-263f-4b48-b775-29911fac6348/recording/TestAccFirebaserulesRelease_BasicRelease.log)]
`TestAccCloudRunService_probes`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/cca5eaa9-263f-4b48-b775-29911fac6348/recording/TestAccCloudRunService_probes.log)]
`TestAccCloudRunService_cloudRunServiceProbesExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/cca5eaa9-263f-4b48-b775-29911fac6348/recording/TestAccCloudRunService_cloudRunServiceProbesExample.log)]
`TestAccSqlDatabaseInstance_mysqlMajorVersionUpgrade`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/cca5eaa9-263f-4b48-b775-29911fac6348/recording/TestAccSqlDatabaseInstance_mysqlMajorVersionUpgrade.log)] 

Tests failed during RECORDING mode:
 `TestAccComputeInstance_soleTenantNodeAffinities`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/cca5eaa9-263f-4b48-b775-29911fac6348/build-log/recording_build/TestAccComputeInstance_soleTenantNodeAffinities_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/cca5eaa9-263f-4b48-b775-29911fac6348/recording/TestAccComputeInstance_soleTenantNodeAffinities.log)]
`TestAccCGCSnippet_eventarcWorkflowsExample`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/cca5eaa9-263f-4b48-b775-29911fac6348/build-log/recording_build/TestAccCGCSnippet_eventarcWorkflowsExample_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/cca5eaa9-263f-4b48-b775-29911fac6348/recording/TestAccCGCSnippet_eventarcWorkflowsExample.log)] 

Please fix these to complete your PR
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/cca5eaa9-263f-4b48-b775-29911fac6348/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/cca5eaa9-263f-4b48-b775-29911fac6348/recording) for each test

rileykarson commented:
Adding @slevenick are reviewer, I haven't had time to address this

slevenick commented on mmv1/products/cloudrun/api.yaml:
```diff
@@ -616,6 +616,79 @@ objects:
+                            description: |-
+                              The header field name.
+                            default_value: ""
+                            send_empty_value: true
```
What does empty mean here? What I want to understand is why a user would want to specify empty rather than some value. Sending an empty header field doesn't make sense, so does empty have a special meaning to the API?

yanweiguo replied:
You're right, empty header name doesn't make sense. I checked what GKE does and GKE doesn't allow empty. Our API needs a change.

slevenick replied:
Should we wait on that change before releasing this then?

yanweiguo replied:
I don't think we need to wait since in Terraform side this field is required.

For header value, IIUC, empty is a valid http header value. Just verified GKE also supports empty value.

slevenick commented on mmv1/products/cloudrun/terraform.yaml:
```diff
@@ -332,6 +332,15 @@ overrides: !ruby/object:Overrides::ResourceOverrides
+        name: "cloud_run_service_probes"
+        min_version: beta
+        primary_resource_id: "default"
+        primary_resource_name: "fmt.Sprintf(\"tf-test-cloudrun-srv%s\", context[\"random_suffix\"])"
```
This is only needed for IAM tests, so can be removed I believe.

yanweiguo replied:
done

slevenick reviewed (commented):
Looks fine overall, had some follow-up questions on the meaning of empty name&value in the headers fields

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-6532-old..auto-pr-6532) ( 1 file changed, 114 insertions(+))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-6532-old..auto-pr-6532) ( 4 files changed, 821 insertions(+))
TF Validator: [Diff](https://github.com/modular-magician/terraform-validator/compare/auto-pr-6532-old..auto-pr-6532) ( 2 files changed, 3 insertions(+), 3 deletions(-))

modular-magician commented:
#### Tests analytics 
Total tests: `2209` 
Passed tests `1964` 
Skipped tests: `240` 
Failed tests: `5` 

#### Action taken 
<details> <summary>Triggering VCR tests in RECORDING mode for the tests that failed during VCR. Click here to see the failed tests</summary><blockquote>TestAccFirebaserulesRelease_BasicRelease|TestAccComputeInstance_soleTenantNodeAffinities|TestAccComputeForwardingRule_internalTcpUdpLbWithMigBackendExample|TestAccCloudRunService_probes|TestAccBigQueryDataTable_bigtable </blockquote></details>

modular-magician commented:
Tests passed during RECORDING mode:
 `TestAccFirebaserulesRelease_BasicRelease`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/08dfa88c-1861-435f-9bbf-68b9fe73f5cf/recording/TestAccFirebaserulesRelease_BasicRelease.log)]
`TestAccComputeForwardingRule_internalTcpUdpLbWithMigBackendExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/08dfa88c-1861-435f-9bbf-68b9fe73f5cf/recording/TestAccComputeForwardingRule_internalTcpUdpLbWithMigBackendExample.log)]
`TestAccCloudRunService_probes`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/08dfa88c-1861-435f-9bbf-68b9fe73f5cf/recording/TestAccCloudRunService_probes.log)]
`TestAccBigQueryDataTable_bigtable`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/08dfa88c-1861-435f-9bbf-68b9fe73f5cf/recording/TestAccBigQueryDataTable_bigtable.log)] 

Tests failed during RECORDING mode:
 `TestAccComputeInstance_soleTenantNodeAffinities`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/08dfa88c-1861-435f-9bbf-68b9fe73f5cf/build-log/recording_build/TestAccComputeInstance_soleTenantNodeAffinities_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/08dfa88c-1861-435f-9bbf-68b9fe73f5cf/recording/TestAccComputeInstance_soleTenantNodeAffinities.log)] 

Please fix these to complete your PR
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/08dfa88c-1861-435f-9bbf-68b9fe73f5cf/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/08dfa88c-1861-435f-9bbf-68b9fe73f5cf/recording) for each test

slevenick commented on mmv1/products/cloudrun/api.yaml:
```diff
@@ -616,6 +616,78 @@ objects:
+                - !ruby/object:Api::Type::NestedObject
+                  name: tcpSocket 
+                  description: |-
+                    TcpSocket speficies an action involving a TCP port.
```
```suggestion
                    TcpSocket specifies an action involving a TCP port.
```

slevenick reviewed (changes requested):
Just one typo change, then should be good to go

EricStG commented:
2 quick questions from an interested party:
- Do you plan to add [GRPC probes](https://cloud.google.com/run/docs/reference/rest/v2/Container#grpcaction) as well?
- since `tcpSocket` and `httpGet` are exclusives, should they include a `conflict` property?

slevenick commented:
> 2 quick questions from an interested party:
> 
> * Do you plan to add [GRPC probes](https://cloud.google.com/run/docs/reference/rest/v2/Container#grpcaction) as well?
> * since `tcpSocket` and `httpGet` are exclusives, should they include a `conflict` property?

Yeah, if that's true they should either have conflicts_with or exactly_one_of on both.

@yanweiguo can you add that if it is appropriate? You can find examples of both in the code base, one thing to note is that you need to hard-code index 0 for nested objects due to how TF represents objects as lists in the underlying structure.

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-6532-old..auto-pr-6532) ( 1 file changed, 114 insertions(+))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-6532-old..auto-pr-6532) ( 4 files changed, 821 insertions(+))
TF Validator: [Diff](https://github.com/modular-magician/terraform-validator/compare/auto-pr-6532-old..auto-pr-6532) ( 2 files changed, 3 insertions(+), 3 deletions(-))

yanweiguo commented:
> 2 quick questions from an interested party:
> 
> * Do you plan to add [GRPC probes](https://cloud.google.com/run/docs/reference/rest/v2/Container#grpcaction) as well?

Yes

> * since `tcpSocket` and `httpGet` are exclusives, should they include a `conflict` property?

Added. PTAL

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-6532-old..auto-pr-6532) ( 1 file changed, 114 insertions(+))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-6532-old..auto-pr-6532) ( 4 files changed, 823 insertions(+))
TF Validator: [Diff](https://github.com/modular-magician/terraform-validator/compare/auto-pr-6532-old..auto-pr-6532) ( 2 files changed, 3 insertions(+), 3 deletions(-))

slevenick commented on mmv1/products/cloudrun/api.yaml:
```diff
@@ -616,6 +616,84 @@ objects:
+                  description: |-
+                    HttpGet specifies the http request to perform.
+                  exactly_one_of:
+                    - tcp_socket
```
This needs to be the entire path to this field like:
template.0.spec.0.containers.0.startup_probe.0.tcp_socket

yanweiguo replied:
done

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-6532-old..auto-pr-6532) ( 1 file changed, 114 insertions(+))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-6532-old..auto-pr-6532) ( 4 files changed, 823 insertions(+))
TF Validator: [Diff](https://github.com/modular-magician/terraform-validator/compare/auto-pr-6532-old..auto-pr-6532) ( 2 files changed, 3 insertions(+), 3 deletions(-))

modular-magician commented:
#### Tests analytics 
Total tests: `2210` 
Passed tests `1966` 
Skipped tests: `240` 
Failed tests: `4` 

#### Action taken 
<details> <summary>Triggering VCR tests in RECORDING mode for the tests that failed during VCR. Click here to see the failed tests</summary><blockquote>TestAccFirebaserulesRelease_BasicRelease|TestAccComputeInstance_soleTenantNodeAffinities|TestAccBigtableAppProfile_bigtableAppProfileAnyclusterExample|TestAccBigtableAppProfile_bigtableAppProfileMulticlusterExample </blockquote></details>

modular-magician commented:
Tests passed during RECORDING mode:
 `TestAccFirebaserulesRelease_BasicRelease`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/9b5f43bb-7eb5-464a-97f3-c7e378294caf/recording/TestAccFirebaserulesRelease_BasicRelease.log)]
`TestAccBigtableAppProfile_bigtableAppProfileAnyclusterExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/9b5f43bb-7eb5-464a-97f3-c7e378294caf/recording/TestAccBigtableAppProfile_bigtableAppProfileAnyclusterExample.log)]
`TestAccBigtableAppProfile_bigtableAppProfileMulticlusterExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/9b5f43bb-7eb5-464a-97f3-c7e378294caf/recording/TestAccBigtableAppProfile_bigtableAppProfileMulticlusterExample.log)] 

Tests failed during RECORDING mode:
 `TestAccComputeInstance_soleTenantNodeAffinities`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/9b5f43bb-7eb5-464a-97f3-c7e378294caf/build-log/recording_build/TestAccComputeInstance_soleTenantNodeAffinities_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/9b5f43bb-7eb5-464a-97f3-c7e378294caf/recording/TestAccComputeInstance_soleTenantNodeAffinities.log)] 

Please fix these to complete your PR
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/9b5f43bb-7eb5-464a-97f3-c7e378294caf/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-6532/artifacts/9b5f43bb-7eb5-464a-97f3-c7e378294caf/recording) for each test
